This is just checking the the data info 

In [2]:
import pandas as pd
csv_path = '/Users/hanfuhou/Downloads/2004to2024all.csv' 
df = pd.read_csv(csv_path)
print("Column Names:")
print(df.columns.tolist())
print("\n Data Types:")
print(df.dtypes)

print("\n  Sample Data (first 5 rows):")
print(df.head())

print("\n  Basic Statistics (numeric columns):")
print(df.describe())

print("\n DataFrame Info:")
df.info()

/var/folders/x7/dd18rcl92fl22lg10gbjt8zm0000gn/T/ipykernel_29297/18498857.py:3: DtypeWarning: Columns (6,13,17) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(csv_path)


Column Names:
['url', 'repository_url', 'labels_url', 'comments_url', 'events_url', 'html_url', 'id', 'node_id', 'number', 'title', 'user', 'labels', 'state', 'locked', 'assignee', 'assignees', 'milestone', 'comments', 'created_at', 'updated_at', 'closed_at', 'author_association', 'type', 'active_lock_reason', 'sub_issues_summary', 'body', 'reactions', 'timeline_url', 'performed_via_github_app', 'state_reason', 'score']

 Data Types:
url                          object
repository_url               object
labels_url                   object
comments_url                 object
events_url                   object
html_url                     object
id                           object
node_id                      object
number                        int64
title                        object
user                         object
labels                       object
state                        object
locked                       object
assignee                     object
assignees             

In [3]:
import pandas as pd


csv_path = '/Users/hanfuhou/Downloads/2004to2024all.csv' 
df = pd.read_csv(csv_path)

# Step 1: Total issues
total_issues = len(df)

# Step 2: Accessibility mentions in labels or title or body
df['labels'] = df['labels'].astype(str).str.lower()
df['title'] = df['title'].astype(str).str.lower()
df['body'] = df['body'].astype(str).str.lower()
accessibility_issues = df[
    df['labels'].str.contains('accessibility|a11y|screen reader|contrast|aria|colorblind|blind|deaf|hearing|ally', na=False) |
    df['title'].str.contains('accessibility|a11y|screen reader|contrast|aria|colorblind|blind|deaf|hearing|ally', na=False) |
    df['body'].str.contains('accessibility|a11y|screen reader|contrast|aria|colorblind|blind|deaf|hearing|ally', na=False)
]
accessibility_count = len(accessibility_issues)
accessibility_ratio = accessibility_count / total_issues

# Step 3: Comments statistics
df['comments_count'] = pd.to_numeric(df['comments'], errors='coerce')
comments_mean = df['comments_count'].mean()
comments_median = df['comments_count'].median()

# Step 4: Author association distribution
author_dist = df['author_association'].value_counts(dropna=False)

# Step 5: Issue states and reasons
state_counts = df['state'].value_counts(dropna=False)
state_reason_counts = df['state_reason'].value_counts(dropna=False)

# Step 6: Score basic stats
score_stats = df['score'].describe()

# === Output ===
print(f"Total issues: {total_issues}")
print(f"Accessibility-related issues: {accessibility_count} ({accessibility_ratio:.2%})")
print(f"Average comments per issue: {comments_mean:.2f}")
print(f"Median comments per issue: {comments_median}")
print("\nAuthor Association Distribution:")
print(author_dist)
print("\nIssue State Counts:")
print(state_counts)
print("\nState Reason Counts:")
print(state_reason_counts)
print("\nScore Stats:")
print(score_stats)

/var/folders/x7/dd18rcl92fl22lg10gbjt8zm0000gn/T/ipykernel_29297/2787716020.py:5: DtypeWarning: Columns (6,13,17) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(csv_path)


Total issues: 404067
Accessibility-related issues: 349340 (86.46%)
Average comments per issue: 4.29
Median comments per issue: 1.0

Author Association Distribution:
author_association
NONE            204096
CONTRIBUTOR      97710
OWNER            45418
MEMBER           32920
COLLABORATOR     23708
MANNEQUIN          213
NaN                  1
1.0                  1
Name: count, dtype: int64

Issue State Counts:
state
closed         294437
open           109628
NaN                 1
CONTRIBUTOR         1
Name: count, dtype: int64

State Reason Counts:
state_reason
completed      285437
NaN            108016
not_planned      8853
reopened         1614
duplicate         147
Name: count, dtype: int64

Score Stats:
count    404065.0
mean          1.0
std           0.0
min           1.0
25%           1.0
50%           1.0
75%           1.0
max           1.0
Name: score, dtype: float64


Author Roles vs. Comments (Issue Engagement)
Average and median comments by author_association
Distribution of comments by author roles
Do OWNER/MEMBER/CONTRIBUTOR issues tend to have more comments?

In [4]:
import pandas as pd
from scipy.stats import spearmanr
import ast

csv_path = '/Users/hanfuhou/Downloads/2004to2024all.csv' 
df = pd.read_csv(csv_path)

# Clean 'author_association'
df['author_association'] = df['author_association'].fillna('NONE')

# Clean 'comments'
df['comments'] = pd.to_numeric(df['comments'], errors='coerce').fillna(0)

# Group by author_association and compute mean comments
author_group = df.groupby('author_association')['comments'].agg(['count', 'mean', 'median']).reset_index()
author_group = author_group.sort_values('mean', ascending=False)

print("\n📊 Author Association vs. Comments (Descriptive Stats):")
print(author_group)

# Correlation between author_association (as ordinal) and comments
mapping = {'NONE':0, 'CONTRIBUTOR':1, 'OWNER':2, 'MEMBER':3, 'COLLABORATOR':4, 'MANNEQUIN':5}
df['author_num'] = df['author_association'].map(mapping).fillna(0)
corr, pval = spearmanr(df['author_num'], df['comments'])

print(f"\n Spearman Correlation (Author Association vs. Comments): {corr:.4f} (p={pval:.4f})")

def count_labels_safe(label_str):
    if pd.isna(label_str):
        return 0
    try:
        if label_str.strip().startswith('['):
            return len(ast.literal_eval(label_str))
        else:
            return len(label_str.split(',')) if ',' in label_str else 1
    except:
        return len(label_str.split(',')) if ',' in label_str else 1

df['label_count'] = df['labels'].apply(count_labels_safe)

# Correlation between label_count and comments
label_corr, label_pval = spearmanr(df['label_count'], df['comments'])

print(f"\n Spearman Correlation (Label Count vs. Comments): {label_corr:.4f} (p={label_pval:.4f})")

# Top labels analysis (optional)
from collections import Counter
import re

label_words = [word.strip().lower() for l in df['labels'].dropna() for word in re.split(r'[,|;]', l)]
label_freq = Counter(label_words)
print("\nTop 10 Labels by Frequency:")
print(label_freq.most_common(10))

/var/folders/x7/dd18rcl92fl22lg10gbjt8zm0000gn/T/ipykernel_29297/1107983517.py:6: DtypeWarning: Columns (6,13,17) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(csv_path)



📊 Author Association vs. Comments (Descriptive Stats):
  author_association   count       mean  median
3          MANNEQUIN     213  10.765258     5.0
4             MEMBER   32920   6.425668     2.0
2        CONTRIBUTOR   97710   4.720612     2.0
5               NONE  204097   4.437277     2.0
1       COLLABORATOR   23708   4.022608     1.0
6              OWNER   45418   1.240411     0.0
0                1.0       1   0.000000     0.0

📈 Spearman Correlation (Author Association vs. Comments): -0.0867 (p=0.0000)

📈 Spearman Correlation (Label Count vs. Comments): 0.3510 (p=0.0000)

Top 10 Labels by Frequency:
[("'default': false", 493499), ('[]', 172114), ("'description': ''}", 89734), ("'description': none}", 79451), ("'default': true", 72096), ("'color': 'ededed'", 63168), ("'description': ''}]", 62442), ("'description': none}]", 56432), ("'name': 'bug'", 34753), ("'name': 'enhancement'", 23522)]


OWNER/MEMBER tend to receive fewer comments.might suggest that issues created by OWNERS/MEMBERS are clearer, easier to resolve, or receive less discussion.
Spearman Correlation: +0.351 shows a positive correlation which means that issues with more labels tend to have more comments.
Many lbels appear to be JSON fragments or API responses, not clean labels like accessibility, bug so I think if we need to get a better insignt on  labels we might need to clean this

Understand whether specific topics in the titles and body text are linked to higher engagement 
TF-IDF Vectorization on Titles + Body
LDA Topic Modeling



In [5]:
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.decomposition import LatentDirichletAllocation
from collections import Counter
from sklearn.preprocessing import LabelEncoder

# === Load Data ===
csv_path = '/Users/hanfuhou/Downloads/2004to2024all.csv' 
df = pd.read_csv(csv_path)

# === Combine Title + Body for Topic Modeling ===
df['text'] = df['title'].fillna('') + ' ' + df['body'].fillna('')

# === TF-IDF Vectorization ===
tfidf = TfidfVectorizer(max_features=5000, stop_words='english')
X_tfidf = tfidf.fit_transform(df['text'])

# === LDA Topic Modeling ===
lda = LatentDirichletAllocation(n_components=5, random_state=42)
topic_probs = lda.fit_transform(X_tfidf)

df['topic'] = np.argmax(topic_probs, axis=1)

# === Average Comments & Stars per Topic ===
df['comment_count'] = pd.to_numeric(df['comments'], errors='coerce').fillna(0)
df['stars'] = pd.to_numeric(df['score'], errors='coerce').fillna(0)

topic_summary = df.groupby('topic').agg({
    'comment_count': ['mean'],
    'stars': ['mean'],
    'author_association': lambda x: dict(Counter(x))
}).reset_index()

print(topic_summary)

# === Top Keywords per Topic ===
feature_names = tfidf.get_feature_names_out()
for idx, topic in enumerate(lda.components_):
    top_words = [feature_names[i] for i in topic.argsort()[-10:][::-1]]
    print(f" Topic {idx} top words:", top_words)

/var/folders/x7/dd18rcl92fl22lg10gbjt8zm0000gn/T/ipykernel_29297/3310591718.py:10: DtypeWarning: Columns (6,13,17) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(csv_path)


  topic comment_count     stars  \
                 mean      mean   
0     0      7.100754  1.000000   
1     1      4.403297  1.000000   
2     2      3.590476  0.999938   
3     3      4.697235  1.000000   
4     4      0.006289  1.000000   

                                  author_association  
                                            <lambda>  
0  {'NONE': 36149, 'CONTRIBUTOR': 4201, 'OWNER': ...  
1  {'CONTRIBUTOR': 6789, 'NONE': 25882, 'OWNER': ...  
2  {'MEMBER': 3308, 'OWNER': 10000, 'NONE': 9831,...  
3  {'COLLABORATOR': 19497, 'OWNER': 19688, 'NONE'...  
4  {'NONE': 48453, 'COLLABORATOR': 9, 'CONTRIBUTO...  
 Topic 0 top words: ['talkback', 'version', 'android', 'google', 'com', 'issue', 'nextcloud', 'app', 'reported', 'original']
 Topic 1 top words: ['td', 'tr', 'java', 'react', 'eslint', 'javascript', 'mozilla', 'autojs', 'version', 'js']
 Topic 2 top words: ['td', 'https', 'tr', 'accessibility', 'com', 'github', 'alt', 'lighthouse', 'li', 'html']
 Topic 3 top words: [

I think for future if we are doing AI benchmarking Topics 0 & 3 could be strong candidates — highly engaged and user-facing.
Topic 4 Almost no comments. Might be ignored issues.